src/client/components/dashboard/NarrativeList/ControlMenu/RevertNarrative.tsx

```diff
@@ -0,0 +1,249 @@
+        if (error instanceof Error) {
+          return error.message;
+        }
+        return 'Unknown error';
+      })();
+
+      this.setState({
+        status: 'error',
+        error: {
+          message,
+        },
+      });
+    }
+  }
+
+  renderError({ error: { message } }: ComponentStateError) {
+    const messageContent = (() => {
+      if (typeof message === 'string') {
+        return <p>{message}</p>;
+      } else {
+        return message.map((message, index) => {
+          return <p key={index}>{message}</p>;
+        });
+      }
+    })();
+    const done = () => {
+      this.props.doneFn();
+      if (this.props.cancelFn) {
+        this.props.cancelFn();
+      }
+    };
+    return (
+      <>
+        <div style={{ fontWeight: 'bold', color: 'red' }}>Error</div>
+        {messageContent}
+        <div style={{ textAlign: 'center' }}>
+          <DashboardButton onClick={done}>Close</DashboardButton>
+        </div>
+      </>
+    );
+  }
+
+  renderLoading(message: string) {
+    return (
+      <div style={{ textAlign: 'center' }}>
+        <LoadingSpinner loading={true} />
+      </div>
+    );
+  }
+
+  renderSuccess() {
+    const done = () => {
+      this.props.doneFn();
+      if (this.props.cancelFn) {
+        this.props.cancelFn();
+      }
+      if (this.props.history) {
+        const { access_group, obj_id } = this.props.narrative;
+        const newUpa = `${access_group}/${obj_id}/${this.state.newVersion}`;
+        const { category } = this.props;
+        const queryParams = new URLSearchParams(location.search);
+        const prefix = '/' + (category === 'own' ? '' : `${category}/`);
+        const newLocation = `${prefix}${newUpa}?${queryParams.toString()}`;
+        this.props.history.push(newLocation);
+      }
+    };
+    return (
+      <div style={{ textAlign: 'center' }}>
+        <p>
+          The Narrative has been successfully reverted to version{' '}
+          {this.props.narrative.version}.
+        </p>
+        <p>
+          It may take up to 30 seconds for this to be reflected in the display.
+        </p>
+        <DashboardButton onClick={done}>Close</DashboardButton>
+      </div>
+    );
+  }
+
+  renderConfirmation() {
+    return (
+      <>
+        <div className="pb2">
+          <p>
+            Reverting a narrative to a previous version is permanent and may
+            cause data loss.
+          </p>
+          <p style={{ fontWeight: 'bold' }}>This action cannot be undone!</p>
+        </div>
+        <div className="pb2">Continue?</div>
+        <div>
+          <DashboardButton
+            onClick={() => this.doRevert()}
+            bgcolor={'red'}
+            textcolor={'white'}
+          >
+            Revert
+          </DashboardButton>
+          <DashboardButton onClick={this.props.cancelFn}>
+            Cancel
+          </DashboardButton>
+        </div>
+      </>
+    );
+  }
+
+  render() {
+    switch (this.state.status) {
```

Comment:
Using `switch` statements can be confusing. They have a lot more functionality that is often unused. This has the result of people not being familiar with them and their features, and they can easily be used incorrectly. For that reason I mostly avoid them, but I recognize that this is a pattern elsewhere in this codebase, so no change is required.